base/meta.jl

```diff
@@ -7,17 +7,66 @@ module Meta
 
 export quot,
        isexpr,
+       propagate_sourceloc!,
        show_sexpr
 
+"""
+    quot(ex)
+
+Return `Expr(:quote, ex)`.  This is useful for interpolating symbols into an
+AST as symbol literals rather than variable names.
+"""
 quot(ex) = Expr(:quote, ex)
 
+"""
+    isexpr(ex, head [, n])
+
+Return true if `ex` is an `Expr` with the given type `head` and optionally that
+the argument list is of length `n`.  `head` may be a `Symbol` or collection of
+`Symbol`s.  For example, to check that a macro was passed a function call
+expression, you might use `isexpr(ex, :call)`.
+"""
 isexpr(ex::Expr, head)          = ex.head === head
 isexpr(ex::Expr, heads::Union{Set,Vector,Tuple}) = in(ex.head, heads)
 isexpr(ex,       head)          = false
 
 isexpr(ex,       head, n::Int)  = isexpr(ex, head) && length(ex.args) == n
 
 
+"""
+    propagate_sourceloc!(location, expr)
+
+Overwrite the caller source location embedded in the macro call AST for all
+macro calls inside `expr`, and return the resulting AST.  This is most useful
+to make the `@__LINE__` and `@__FILE__` macros return the location where a
+parent macro was invoked.  For example:
+
+```
+macro example()
+    propagate_sourceloc!(__source__,
+        quote
+            @__LINE__
```

Comment:
I think I'm not understanding the purpose of this example, so I'm wary of, erhm, propagating it... If it's trying to solve https://github.com/c42f/MicroLogging.jl/commit/f6db3715c681af90511f5219252f16e57096a194#commitcomment-20846418 then why not interpolate `__source__.line`? `@__LINE__` is documented and implemented as a literal (that's why we need the parser to insert it..), so I think this would be confusing.